In [51]:
import pandas as pd

df_restaurants = pd.read_csv('data/raw/restaurants.csv')
df_menus = pd.read_csv('data/raw/restaurant-menus.csv')


# Convert Price to float

In [52]:
df_menus.info()
df_restaurants.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5117217 entries, 0 to 5117216
Data columns (total 5 columns):
 #   Column         Dtype 
---  ------         ----- 
 0   restaurant_id  int64 
 1   category       object
 2   name           object
 3   description    object
 4   price          object
dtypes: int64(1), object(4)
memory usage: 195.2+ MB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 63469 entries, 0 to 63468
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   id            63469 non-null  int64  
 1   position      63469 non-null  int64  
 2   name          63469 non-null  object 
 3   score         35302 non-null  float64
 4   ratings       35302 non-null  float64
 5   category      63384 non-null  object 
 6   price_range   52852 non-null  object 
 7   full_address  63016 non-null  object 
 8   zip_code      62952 non-null  object 
 9   lat           63469 non-null  float64
 10  lng           634

In [53]:
df_menus['price'] = df_menus['price'].apply(lambda x: float(x[:-4]))
df_menus.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5117217 entries, 0 to 5117216
Data columns (total 5 columns):
 #   Column         Dtype  
---  ------         -----  
 0   restaurant_id  int64  
 1   category       object 
 2   name           object 
 3   description    object 
 4   price          float64
dtypes: float64(1), int64(1), object(3)
memory usage: 195.2+ MB


## Extract Average Price and Price Sum and min and max price Per Restaurant

In [54]:
df_price_mean_sum = df_menus.groupby('restaurant_id')['price'].mean().reset_index()
df_price_sum = df_menus.groupby('restaurant_id')['price'].sum()
df_price_min = df_menus.groupby('restaurant_id')['price'].min()
df_price_max = df_menus.groupby('restaurant_id')['price'].max()

df_price_mean_sum['price_sum'] = df_price_sum.values
df_price_mean_sum['price_min'] = df_price_min.values
df_price_mean_sum['price_max'] = df_price_max.values
df_price_mean_sum.rename(columns={'price': 'avg_price'}, inplace=True)

df_price_mean_sum.info()
df_price_mean_sum

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 63309 entries, 0 to 63308
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   restaurant_id  63309 non-null  int64  
 1   avg_price      63309 non-null  float64
 2   price_sum      63309 non-null  float64
 3   price_min      63309 non-null  float64
 4   price_max      63309 non-null  float64
dtypes: float64(4), int64(1)
memory usage: 2.4 MB


restaurant_id  avg_price  price_sum  price_min  price_max
0                  1   5.663684     538.05       1.89      16.99
1                  2   5.505333     330.32       0.60       8.49
2                  3  10.762143     301.34       3.00      16.78
3                  4  10.531892     389.68       2.49      16.99
4                  5   4.532576     299.15       1.00      20.00
...              ...        ...        ...        ...        ...
63304          63465   7.680833    1106.04       0.00      29.99
63305          63466   5.977500      23.91       4.73       9.46
63306          63467   7.625000      30.50       4.73      10.52
63307          63468  14.349130     660.06       0.00      44.99
63308          63469   5.006219    1006.25       0.75      27.48

[63309 rows x 5 columns]

## join the price data with the restaurant data

In [55]:
merged_df = df_restaurants.merge(df_price_mean_sum, left_on='id', right_on='restaurant_id', how='left')
merged_df.info()
merged_df

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 63469 entries, 0 to 63468
Data columns (total 16 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             63469 non-null  int64  
 1   position       63469 non-null  int64  
 2   name           63469 non-null  object 
 3   score          35302 non-null  float64
 4   ratings        35302 non-null  float64
 5   category       63384 non-null  object 
 6   price_range    52852 non-null  object 
 7   full_address   63016 non-null  object 
 8   zip_code       62952 non-null  object 
 9   lat            63469 non-null  float64
 10  lng            63469 non-null  float64
 11  restaurant_id  63309 non-null  float64
 12  avg_price      63309 non-null  float64
 13  price_sum      63309 non-null  float64
 14  price_min      63309 non-null  float64
 15  price_max      63309 non-null  float64
dtypes: float64(9), int64(2), object(5)
memory usage: 7.7+ MB


id  position                                            name  score  \
0          1        19               PJ Fresh (224 Daniel Payne Drive)    NaN   
1          2         9                  J' ti`'z Smoothie-N-Coffee Bar    NaN   
2          3         6  Philly Fresh Cheesesteaks (541-B Graymont Ave)    NaN   
3          4        17         Papa Murphy's (1580 Montgomery Highway)    NaN   
4          5       162                Nelson Brothers Cafe (17th St N)    4.7   
...      ...       ...                                             ...    ...   
63464  63465         2                           IHOP (4321 N 10th St)    4.6   
63465  63466        29                       Feldman's - N 10th Street    NaN   
63466  63467        27                           Feldman's - N 23rd St    NaN   
63467  63468        19           Church's Chicken (620 S. 10th Street)    4.3   
63468  63469         9                       7-Eleven (1600 S 23rd St)    NaN   

       ratings                                           category price_range  \
0          NaN                      Burgers, American, Sandwiches           $   
1          NaN   Coffee and Tea, Breakfast and Brunch, Bubble Tea         NaN   
2          NaN         American, Cheesesteak, Sandwiches, Alcohol           $   
3          NaN                                              Pizza           $   
4         22.0          Breakfast and Brunch, Burgers, Sandwiches         NaN   
...        ...                                                ...         ...   
63464     56.0  American, Breakfast and Brunch, burger, Comfor...           $   
63465      NaN                                                NaN         NaN   
63466      NaN                                                NaN         NaN   
63467     17.0                            Fast Food, Family Meals           $   
63468      NaN  Everyday Essentials, Convenience, Home &amp; P...           $   

                                            full_address zip_code        lat  \
0          224 Daniel Payne Drive, Birmingham, AL, 35207    35207  33.562365   
1      1521 Pinson Valley Parkway, Birmingham, AL, 35217    35217  33.583640   
2              541-B Graymont Ave, Birmingham, AL, 35204    35204  33.509800   
3             1580 Montgomery Highway, Hoover, AL, 35226    35226  33.404439   
4                   314 17th St N, Birmingham, AL, 35203    35203  33.514730   
...                                                  ...      ...        ...   
63464                4321 N 10th St, MC ALLEN, TX, 78504    78504  26.242853   
63465                 3010 N 10th Street, Mc Allen, TX,       NaN  26.230764   
63466              6101 N 23rd Street #B, Mc Allen, TX,       NaN  26.260353   
63467             620 S. 10th Street, McAllen, TX, 78501    78501  26.197333   
63468                 1600 S 23rd St, Mcallen, TX, 78503    78503  26.191710   

             lng  restaurant_id  avg_price  price_sum  price_min  price_max  
0     -86.830703            1.0   5.663684     538.05       1.89      16.99  
1     -86.773330            2.0   5.505333     330.32       0.60       8.49  
2     -86.854640            3.0  10.762143     301.34       3.00      16.78  
3     -86.806614            4.0  10.531892     389.68       2.49      16.99  
4     -86.811700            5.0   4.532576     299.15       1.00      20.00  
...          ...            ...        ...        ...        ...        ...  
63464 -98.223984        63465.0   7.680833    1106.04       0.00      29.99  
63465 -98.224922        63466.0   5.977500      23.91       4.73       9.46  
63466 -98.236755        63467.0   7.625000      30.50       4.73      10.52  
63467 -98.231369        63468.0  14.349130     660.06       0.00      44.99  
63468 -98.247880        63469.0   5.006219    1006.25       0.75      27.48  

[63469 rows x 16 columns]

### Check for missing values for merged_df

In [56]:
merged_df[['price_max', 'price_min', 'price_sum', 'zip_code']].isna().sum()

price_max    160
price_min    160
price_sum    160
zip_code     517
dtype: int64

### there are 160 record missing for price_max, price_min and price_sum out of 63469
### so it won't make problem if remove 160 records

In [57]:
merged_df.dropna(subset=['price_max', 'price_min', 'price_sum', 'zip_code'], inplace=True)
merged_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 62792 entries, 0 to 63468
Data columns (total 16 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             62792 non-null  int64  
 1   position       62792 non-null  int64  
 2   name           62792 non-null  object 
 3   score          35018 non-null  float64
 4   ratings        35018 non-null  float64
 5   category       62770 non-null  object 
 6   price_range    52426 non-null  object 
 7   full_address   62792 non-null  object 
 8   zip_code       62792 non-null  object 
 9   lat            62792 non-null  float64
 10  lng            62792 non-null  float64
 11  restaurant_id  62792 non-null  float64
 12  avg_price      62792 non-null  float64
 13  price_sum      62792 non-null  float64
 14  price_min      62792 non-null  float64
 15  price_max      62792 non-null  float64
dtypes: float64(9), int64(2), object(5)
memory usage: 8.1+ MB


## get data frame without price range with null values to calculate max and min price for each category

In [58]:
price_range_analysis = merged_df.dropna(subset=['price_range'])
price_range_analysis.groupby('price_range')[['price_max', 'price_min']].mean().reset_index()

price_range  price_max  price_min
0                  $  34.352069   1.547300
1                 $$  35.316274   2.434294
2                $$$  60.360901   3.352532
3               $$$$  95.721200   6.206400
4  $$$$$$$$$$$$$$$$$   5.490000   0.000000

## fill missing values for price range using min and max price that we got from previous step

In [59]:
def get_price_range(mn, mx):
    if mx <= 34.352069 and mn >= 0:
        return '$'
    elif mx <= 35.316274 and mn >= 2.434294:
        return '$$'
    elif mx <= 60.360901 and mn >= 3.352532:
        return '$$$'
    else:
        return '$$$$'


merged_df.loc[merged_df['price_range'].isna(), 'price_range'] = \
    merged_df[merged_df['price_range'].isna()].apply(lambda row: get_price_range(row['price_min'], row['price_max']),
                                                     axis=1)
merged_df.drop_duplicates()
merged_df.info()


<class 'pandas.core.frame.DataFrame'>
Index: 62792 entries, 0 to 63468
Data columns (total 16 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             62792 non-null  int64  
 1   position       62792 non-null  int64  
 2   name           62792 non-null  object 
 3   score          35018 non-null  float64
 4   ratings        35018 non-null  float64
 5   category       62770 non-null  object 
 6   price_range    62792 non-null  object 
 7   full_address   62792 non-null  object 
 8   zip_code       62792 non-null  object 
 9   lat            62792 non-null  float64
 10  lng            62792 non-null  float64
 11  restaurant_id  62792 non-null  float64
 12  avg_price      62792 non-null  float64
 13  price_sum      62792 non-null  float64
 14  price_min      62792 non-null  float64
 15  price_max      62792 non-null  float64
dtypes: float64(9), int64(2), object(5)
memory usage: 8.1+ MB


### save the processed data in `data/processed/processed_restaurants.csv`

In [60]:
merged_df.to_csv('data/processed/processed_restaurants.csv', index=False)